src/algebra/ring/basic.lean

```diff
@@ -356,6 +356,20 @@ variables {rα : non_assoc_semiring α} {rβ : non_assoc_semiring β}
 
 include rα rβ
 
+instance : add_monoid_hom_class (α →+* β) α β :=
+{ coe := ring_hom.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_add := ring_hom.map_add',
+  map_zero := ring_hom.map_zero' }
+
+instance : monoid_hom_class (α →+* β) α β :=
+{ coe := ring_hom.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_mul := ring_hom.map_mul',
+  map_one := ring_hom.map_one' }
```

Comment:
Would it be better here to go straight to adding a `ring_hom_class`? Or would you argue there is no point in such a class, and using these two classes separately is easier?

Reply:
With the current design we'd need a separate `ring_hom_class`, since taking the `add_monoid_hom_class` and `monoid_hom_class` in separate parameters can have different `to_fun`s.

An alternative is to unbundle the `fun_like` from the `_hom_class`es, taking it as an instance parameter rather than extending it:
```lean
class zero_hom_class (F : Type*) (M N : out_param $ Type*) [fun_like F M (λ _, N)]
  [has_zero M] [has_zero N] :=
(map_zero : ∀ (f : F), f 0 = 0)
```
Then we wouldn't need to define the conjunction of the `_hom_class`es separately, just one per operation. The disadvantage being that the list of instance parameters becomes extremely long:
```lean
theorem polynomial.hom_eval₂ {R S T : Type*} [semiring R] (p : polynomial R)
  [comm_semiring S] [comm_semiring T]
  (f : R →+* S) (g : S →+* T) (x : S) :
  ⇑g (polynomial.eval₂ f x p) = polynomial.eval₂ (g.comp f) (⇑g x) p
```
becomes
```lean
theorem polynomial.hom_eval₂ {R S T F G : Type*} [semiring R] (p : polynomial R)
  [comm_semiring S] [comm_semiring T]
  [fun_like F R (λ _, S)] [add_hom_class F R S] [mul_hom_class F R S] [one_hom_class F R S] [zero_hom_class F R S]
  [fun_like G S (λ _, T)] [add_hom_class G S T] [mul_hom_class G S T] [one_hom_class G S T] [zero_hom_class G S T]
  (f : F) (g : G) (x : S) :
  ⇑g (polynomial.eval₂ f x p) = polynomial.eval₂ (g.comp f) (⇑g x) p
```
instead of my preferred bundled design:
```lean
theorem polynomial.hom_eval₂ {R S T F G : Type*} [semiring R] (p : polynomial R)
  [comm_semiring S] [comm_semiring T] [ring_hom_class F R S] [ring_hom_class G S T]
  (f : F) (g : G) (x : S) :
  ⇑g (polynomial.eval₂ f x p) = polynomial.eval₂ (g.comp f) (⇑g x) p
```

Reply:
Note that if we just want to shrink the argument list, we can still have a vestigial `ring_hom_class` that extends the others and use `[fun_like G S (λ _, T)] [ring_hom_class G S T]`; which _maybe_ has an advantage since there are no longer any diamonds in the data-carrying classes. I think your design is fine though.

In that case, can you replace these two instances with `class ring_hom_class` and `instance : ring_hom_class`?

Reply:
Indeed, I think we'll end up having to unbundle parts of the linear/mul_action hom classes anyway: if it was a subclass of `fun_like` we'd get a dangerous instance `linear_map_class F σ M N → fun_like F M N`. But since they're also `add_hom`s, the class would look something more like:
```lean
class linear_map_class (F : Type*) {R S : Type*} [semiring R] [semiring S] (σ : R →+* S) (M M₂ : Type*)
  [add_comm_monoid M] [add_comm_monoid M₂] [module R M] [module S M₂]
  [add_hom_class F M M₂] := -- alternatively [fun_like F M (λ _, M₂)] [add_hom_class F M M₂] :=
(map_smul' : ∀ (f : F) (r : R) (x : M), f (r • x) = (σ r) • f x)
```
so unbundling `fun_like` from the `add_hom_class` doesn't really bring anything in that case.

Reply:
I've added a commit defining a `ring_hom_class`. I expect nothing will break, let's see :P